include/zephyr/drivers/fuel_gauge.h

```diff
@@ -156,6 +165,16 @@ struct fuel_gauge_get_property {
 		uint16_t sbs_at_rate_time_to_empty;
 		/** FUEL_GAUGE_SBS_ATRATE_OK */
 		bool sbs_at_rate_ok;
+		/** FUEL_GAUGE_MANUFACTURE_DATE */
+		uint16_t manufacturer_date;
+		/** FUEL_GAUGE_SERIAL_NUMBER */
+		uint16_t serial_number;
+		/** FUEL_GAUGE_MANUFACTURER_NAME */
```

Comment:
So the direction we wanted to go with the API was adding another function that's solely for larger array-backed properties.

I'm wary to extend the existing API like this because it's potentially easy to use it without setting the pointer and cause hard to debug memory corruption bugs. Sure that'd be the client's fault, but I think that'd be a common enough mistake that I'd like to protect against it.

The direction we were heading was to add a function like so: 

```
__syscall int fuel_gauge_get_big_prop(const struct device *dev, struct fuel_gauge_big_property *p, size_t dst_len, void *dst);
```

or if we forego the whole struct parameter:

How would you feel about this change to the API? Would you be okay changing your PR to reflect a design like that?

@teburd do you have any thoughts here?

Reply:
I do check for a non-NULL pointer before using the buffer, but get your point.  Even if someone assigns a buffer to the pointer in the union, there's really no guarantee that they've assigned a large enough buffer, whereas the above call, since you're passing a size, kind of ensures that the buffer should work - or at least takes a little guesswork out of the call.  
Only downside is that the user has to know to call 'get big' for the string properties, but that seems like a reasonable tradeoff.  Let me rework the PR to include this.

Reply:
Alternately, what if we just added the size to the fuel_gauge_get_property structure?  Along with checking the pointer, can check that the size is non-zero.  The likelyhood of trouble using the function should be minimized in this way, and string properties could still be retrieved using the same API as other registers.

Still would have the opportunity to specify a buffer, and the wrong size, but even with the new API, could do that, and cause memory corruption issues.

Structure would look like below:
		struct fuel_gauge_get_property props[] = {
			{
				.property_type = FUEL_GAUGE_VOLTAGE,
			},
			{
				.property_type = FUEL_GAUGE_DEVICE_CHEMISTRY,
				.value.device_chemistry = (char *)chem_buffer,
				.value.buf_len = sizeof(chem_buffer),
			},
		};

Reply:
> I  do check for a non-NULL pointer before using the buffer

We're also not guaranteed that the pointer will be NULL if they haven't set it. That's what really scares me as well as the length of the buffer as you mentioned. Memory errors are scary.

> Still would have the opportunity to specify a buffer, and the wrong size, but even with the new API, could do that, and cause memory corruption issues.

Similarly I'm concerned that if we do this a client may forget to set one of these fields. Unfortunately we can't assume these values will be initialized to zero/NULL. If we could assume such initialization, I'd be 100% agreement for what you just described.

@teburd What are your thoughts here? I tend to err on the side of caution.

Reply:
@aaronemassey - updated to make the string properties separate from the "regular" registers.  I'm allocating the memory in the union, so should be safe from a memory perspective.  Thoughts?

Reply:
Apologies for keeping you waiting. Was sick yesterday.

> should be safe from a memory perspective

This is definitely memory safe.

So I only see two problems with the API as it stands now.
1) It's pretty specific to NULL terminated strings instead of arbitrary data that we do want to support.
2) No matter what string/big property I want, I have to allocate the max size of bytes on my stack. It's the drawback of a union.

`1.` may be solved easy enough by adding a length field to our new struct here that gets set by the driver and read by the client.

`2.` is a bit trickier. For the properties we have now, I don't think we care about allocating 21 bytes even if we're just grabbing 10 bytes or so. However, there may come a day that we have a 64 byte property to support. We certainly have up to 32 byte properties to support already for SBS. I'd rather not require a client to allocate 32 or 64 bytes on the stack per struct allocated regardless if they're just getting a 10 byte value.

How would you feel if we went with the approach I mentioned above:

```
__syscall int fuel_gauge_get_big_prop(const struct device *dev, struct fuel_gauge_big_property *p, size_t dst_len, void *dst);

```

Where we let clients decide how much memory to give us with the `dst_len` and `dst` parameters which may be informed by constants defined in the `fuel_gauge.h` header like `SBS_CHEMISTRY_MAX_LEN`.

However, this version I just described has one obvious drawback: We have to make multiple calls to get each parameter. This can get significant if syscalls are enabled because we're potentially making multiple syscalls. But I don't think that's a huge price to pay given most of these string/array-backed properties are typically retrieved only once and then forever cached. I also think since these tend to be larger properties with many multiple bytes to read, there's less gain from optimizing burst reads on the driver level like we discussed on Discord; let me know if you think that's not the case.

What do you think @protterc ?

Reply:
Hope you're feeling better.  

Good point about the stack, it could get quite large for the one call where you're reading string properties, and could cause apps to have to oversize the task stack, just for this one call to get the static string properties.

I'm thinking of changing the name, not to 'big', but to 'block' - that's how they're referenced in the SBS spec, and it seems to fit well.  They do also reference them as strings, but as you pointed out, even though they're not typical null-terminated C strings, we should probably not overload that nomenclature.

So, I'm going to use your API above, where the app will allocate the memory, and pass it down to the driver.  The struct won't have a 'value' field, just the status, etc.  One nice side effect of creating the call this way is that I can use the 'block' call to also read all the dynamic registers in one shot - don't have to add that logic to the driver to decide if they're all sequential.

Reply:
Updated to the requested API